### ISSUE TITLE
Delivered phases should be on top of all phases, and update the delivery date as the final date.

### ISSUE THREAD

**Issue by vic-topcoder** [appirio-tech/connect-app#2183]
Right now when we change the status of a phase to Delivered, we don't change the delivery date, and it should stay on top of all active/planned phases.

The "Delivered" state really represent when we close the work and is used to recognize revenue. It should update the end date to the date of selecting the status and thus move to the top of the active planned phases (Why do we have planned phases in the past??)
![screen shot 2018-07-10 at 7 04 13 pm](https://user-images.githubusercontent.com/8484061/42546896-5034ae58-8474-11e8-82d7-cfe86d327aa6.png)

**Comment by vikasrohit** [comment]
> (Why do we have planned phases in the past??)

As I said previously, we are not forcing any thing to the user. So, its their will how they want to organize the phases. If they have marked a phase completed but haven't marked a phase previous to it, it is their responsibility to update the dates, of the previous phase, to be ahead of the completed one.

What we can do is to update the end date of the phase if it is closed earlier than scheduled date.

fyi @rmjmccormick

**Comment by rmjmccormick** [comment]
@vikasrohit Thanks. As you said, at this stage, we are giving the user full flexibility and responsibility. @vic-topcoder 's point about showing delivered phases at the top of the list of phases (above any other active or planned phases) is a good one though, can this be done?

**Comment by vikasrohit** [comment]
Yes, it can be done @rmjmccormick but it would introduce a custom sorting logic after sorting from server which I like to avoid to keep things simple. I think these type of use cases would be rare if users are using the platform correctly. I like to avoid such cases because they tend to increase possibility of bugs around that feature.
However, its your final call, we can make those changes.

**Comment by vic-topcoder** [comment]
Ideally we should record the "Delivery date" as a separate field so for our intended purposes we have a planned start date, actual start date, planned delivery date, actual delivery date, and we store those so we can analyze them later. Also - we should record any extensions to the project phase as to know why and what happened.

On another note, Tony requested that we keep a history log of all time changes/planning/execution on the phase for posterity, so we might think about how we keep a record of all activity on the project plan.

**Comment by vikasrohit** [comment]
@vic-topcoder Yes, we thought of keeping a separate "Delivery date", however, when we did brainstorming for milestones, we found that we can utilize the actual end date (which are keeping as a separate field in each milestone as completion date) to determine when the phase was actually delivered. Does that serve your purpose?
For audit thing, we have that in mind since when we migrated our message api from discourse. https://github.com/appirio-tech/tc-message-service/issues/61. I tried get opinion of team in slack as well, but I think it didn't catch the required attention at that time. We have a plan for implementing generic audit feature across apps in our system which we can kick off if have it now on priority and have the budget.

**Comment by gondzo** [comment]
Updating end date is handled in #2187 - should we close this one @vikasrohit
